File: backend/api/routes/updates.py
"""Update manager API routes (P4-16, P4-21d).

Endpoints for checking database updates, triggering updates,
viewing update history, managing re-annotation prompts,
and checking for app updates via GitHub Releases API.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from backend.db.connection import get_registry
from backend.db.database_registry import DATABASES, get_build_fn, get_database_status
from backend.db.update_manager import (
    AUTO_UPDATE_DEFAULTS,
    check_all_updates,
    dismiss_prompt,
    format_version_display,
    get_active_prompts,
    get_all_version_stamps,
    get_auto_update,
    get_update_history,
    set_auto_update,
    should_download_now,
)

router = APIRouter(prefix="/updates", tags=["updates"])


# ── Response models ──────────────────────────────────────────────────


class UpdateAvailable(BaseModel):
    db_name: str
    latest_version: str
    download_size_bytes: int
    release_date: str | None = None


class UpdateCheckResponse(BaseModel):
    available: list[UpdateAvailable]
    up_to_date: list[str]
    errors: list[str]
    checked_at: str


class UpdateHistoryEntry(BaseModel):
    id: int
    db_name: str
    previous_version: str | None
    new_version: str
    updated_at: str | None
    variants_added: int | None
    variants_reclassified: int | None
    download_size_bytes: int | None
    duration_seconds: int | None


class WatchedReclassification(BaseModel):
    rsid: str
    gene_symbol: str | None = None
    old_significance: str
    new_significance: str


class ReannotationPrompt(BaseModel):
    id: int
    sample_id: int
    db_name: str
    db_version: str
    candidate_count: int
    watched_count: int = 0
    watched_details: list[WatchedReclassification] = []
    created_at: str | None


# Finding-level change diff (SW-A4b / #8) — the deferred second half of SW-A4.


class FindingFieldChange(BaseModel):
    field: str
    before: str | None = None
    after: str | None = None


class ReleaseDelta(BaseModel):
    db_name: str
    before: str | None = None
    after: str | None = None


class ChangedFinding(BaseModel):
    module: str
    category: str | None = None
    gene_symbol: str | None = None
    rsid: str | None = None
    drug: str | None = None
    diplotype: str | None = None
    finding_text: str
    changes: list[FindingFieldChange] = []


class DiffFinding(BaseModel):
    module: str
    category: str | None = None
    gene_symbol: str | None = None
    rsid: str | None = None
    drug: str | None = None
    diplotype: str | None = None
    finding_text: str
    clinvar_significance: str | None = None
    evidence_level: int | None = None
    metabolizer_status: str | None = None
    pathway_level: str | None = None


class FindingChangesResponse(BaseModel):
    """What changed at the finding level since the prior analysis.

    ``available`` is False when there is no undismissed diff with changes — the
    common case (first analysis, nothing changed, or the user dismissed it).
    """

    available: bool
    generated_at: str | None = None
    release_deltas: list[ReleaseDelta] = []
    changed: list[ChangedFinding] = []
    added: list[DiffFinding] = []
    removed: list[DiffFinding] = []
    counts: dict[str, int] = {}


class DatabaseStatus(BaseModel):
    db_name: str
    display_name: str
    current_version: str | None
    version_display: str | None
    downloaded_at: str | None
    file_size_bytes: int | None
    auto_update: bool
    update_available: bool


class TriggerUpdateRequest(BaseModel):
    db_name: str


class TriggerUpdateResponse(BaseModel):
    job_id: str
    db_name: str
    message: str


class AutoUpdateRequest(BaseModel):
    db_name: str
    enabled: bool


class AutoUpdateResponse(BaseModel):
    db_name: str
    enabled: bool


# ── Endpoints ────────────────────────────────────────────────────────


@router.get("/check", response_model=UpdateCheckResponse)
async def check_for_updates() -> UpdateCheckResponse:
    """Check all databases for available updates."""
    registry = get_registry()
    result = check_all_updates(registry.reference_engine, settings=registry.settings)

    return UpdateCheckResponse(
        available=[
            UpdateAvailable(
                db_name=v.db_name,
                latest_version=v.latest_version,
                download_size_bytes=v.download_size_bytes,
                release_date=v.release_date,
            )
            for v in result.available
        ],
        up_to_date=result.up_to_date,
        errors=result.errors,
        checked_at=result.checked_at.isoformat(),
    )


@router.post("/trigger", response_model=TriggerUpdateResponse, status_code=202)
async def trigger_update(req: TriggerUpdateRequest) -> TriggerUpdateResponse:
    """Trigger an update for a specific database.

    Enqueues the update as a background Huey task and returns
    the job_id for progress tracking via SSE.
    """
    from backend.db.update_manager import _BUNDLE_DBS
    from backend.tasks.huey_tasks import (
        create_database_update_job,
        run_database_update_task,
    )

    # Any database with a build function, or a manifest-driven bundle
    # (vep_bundle / lai_bundle / ancestry_pca), can be updated. The bundle set
    # must mirror the scheduler's _dispatch_auto_update and huey's
    # run_database_update_task so the UI never offers an update the backend
    # rejects (previously lai_bundle/ancestry_pca 400'd here).
    db_info = DATABASES.get(req.db_name)
    build_fn = get_build_fn(req.db_name) if db_info else None
    is_bundle = req.db_name in _BUNDLE_DBS
    if build_fn is None and not is_bundle:
        supported = sorted(
            {k for k in DATABASES if get_build_fn(k) is not None} | set(_BUNDLE_DBS)
        )
        raise HTTPException(
            status_code=400,
            detail=f"Update not supported for '{req.db_name}'. Supported: {supported}",
        )

    # Check bandwidth window — look up actual expected download size
    registry = get_registry()
    settings = registry.settings
    estimated_size = db_info.expected_size_bytes if db_info else 0
    if not should_download_now(estimated_size, settings.update_download_window):
        raise HTTPException(
            status_code=409,
            detail=f"Outside bandwidth window ({settings.update_download_window}).",
        )

    job_id = create_database_update_job(req.db_name)
    run_database_update_task(job_id, req.db_name)

    return TriggerUpdateResponse(
        job_id=job_id,
        db_name=req.db_name,
        message=f"Update queued for {req.db_name}",
    )


class JobStatusResponse(BaseModel):
    job_id: str
    status: str
    progress_pct: float
    message: str
    error: str | None = None


@router.get("/job/{job_id}", response_model=JobStatusResponse)
async def get_update_job_status(job_id: str) -> JobStatusResponse:
    """Poll the status of a database update job."""
    from backend.api.sse import get_job_progress

    registry = get_registry()
    status = get_job_progress(registry.reference_engine, job_id)
    if status is None:
        raise HTTPException(status_code=404, detail="Job not found")
    return JobStatusResponse(
        job_id=status.job_id,
        status=status.status,
        progress_pct=status.progress_pct,
        message=status.message,
        error=status.error,
    )


@router.get("/history", response_model=list[UpdateHistoryEntry])
async def list_update_history(
    db_name: str | None = None,
    limit: int = Query(default=50, ge=1, le=500),
) -> list[UpdateHistoryEntry]:
    """Get update history, optionally filtered by database name."""
    registry = get_registry()
    rows = get_update_history(registry.reference_engine, db_name=db_name, limit=limit)
    return [UpdateHistoryEntry(**r) for r in rows]


@router.get("/status", response_model=list[DatabaseStatus])
async def get_database_statuses() -> list[DatabaseStatus]:
    """Get current version and display info for all tracked databases.

    Note: ``update_available`` is always False here. Call
    ``GET /updates/check`` for actual update availability (network call).
    """
    registry = get_registry()
    engine = registry.reference_engine
    settings = registry.settings

    # Fetch all version stamps in one query
    stamps = {s["db_name"]: s for s in get_all_version_stamps(engine)}

    # Check which databases have updates available (cached, no network call)
    # We only mark update_available based on whether we have a version at all;
    # actual update checks are done via GET /updates/check
    result = []
    for db_name in AUTO_UPDATE_DEFAULTS:
        auto_update = get_auto_update(engine, db_name)
        stamp = stamps.get(db_name)
        version = stamp["version"] if stamp else None
        downloaded_at = stamp["downloaded_at"] if stamp else None
        file_size_bytes = stamp["file_size_bytes"] if stamp else None
        db_info = DATABASES.get(db_name)
        display_name = db_info.display_name if db_info else db_name

        # If no version stamp or file size, check on-disk status as fallback.
        # Databases may exist on disk without a database_versions entry
        # (e.g. after a build that didn't record its version).
        on_disk = None
        if (version is None or file_size_bytes is None) and db_info is not None:
            on_disk = get_database_status(db_info, settings)

        if version is None and on_disk is not None and on_disk["downloaded"]:
            version = "installed"
            if on_disk.get("file_size_bytes"):
                file_size_bytes = on_disk["file_size_bytes"]

        if file_size_bytes is None and on_disk is not None and on_disk.get("file_size_bytes"):
            file_size_bytes = on_disk["file_size_bytes"]

        result.append(
            DatabaseStatus(
                db_name=db_name,
                display_name=display_name,
                current_version=version,
                version_display=format_version_display(version, db_name),
                downloaded_at=downloaded_at,
                file_size_bytes=file_size_bytes,
                auto_update=auto_update,
                update_available=False,  # Set by GET /updates/check
            )
        )
    return result


@router.get("/prompts", response_model=list[ReannotationPrompt])
async def list_reannotation_prompts(
    sample_id: int | None = None,
) -> list[ReannotationPrompt]:
    """Get active (undismissed) re-annotation prompts."""
    registry = get_registry()
    rows = get_active_prompts(registry.reference_engine, sample_id=sample_id)
    return [ReannotationPrompt(**r) for r in rows]


class AppUpdateResponse(BaseModel):
    update_available: bool
    current_version: str
    latest_version: str | None = None
    release_url: str | None = None
    release_notes: str | None = None
    error: str | None = None


@router.get("/app-update", response_model=AppUpdateResponse)
async def check_app_update() -> AppUpdateResponse:
    """Check GitHub Releases API for a newer Yeliztli version."""
    from backend.utils.update_checker import check_app_update as _check

    info = await _check()
    return AppUpdateResponse(
        update_available=info.update_available,
        current_version=info.current_version,
        latest_version=info.latest_version,
        release_url=info.release_url,
        release_notes=info.release_notes,
        error=info.error,
    )


@router.post("/prompts/{prompt_id}/dismiss")
async def dismiss_reannotation_prompt(prompt_id: int) -> dict:
    """Dismiss a re-annotation prompt."""
    registry = get_registry()
    ok = dismiss_prompt(registry.reference_engine, prompt_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Prompt not found")
    return {"status": "dismissed", "prompt_id": prompt_id}


@router.get("/finding-changes", response_model=FindingChangesResponse)
async def get_finding_changes(
    sample_id: int = Query(..., description="Sample ID"),
) -> FindingChangesResponse:
    """Return the finding-level change diff for a sample's latest re-annotation.

    Disclosure only (SW-A4b): reports which findings were added / removed /
    meaning-shifted since the prior analysis and the source-release delta that
    explains it. Responds ``available=False`` when there is no undismissed diff
    with changes.
    """
    from backend.analysis.finding_diff import has_changes, read_finding_diff
    from backend.api.routes.risk_common import resolve_sample_engine

    engine = resolve_sample_engine(sample_id)
    diff = read_finding_diff(engine)
    if diff is None or diff.get("dismissed") or not has_changes(diff):
        return FindingChangesResponse(available=False)

    return FindingChangesResponse(
        available=True,
        generated_at=diff.get("generated_at"),
        release_deltas=[ReleaseDelta(**d) for d in diff.get("release_deltas", [])],
        changed=[ChangedFinding(**c) for c in diff.get("changed", [])],
        added=[DiffFinding(**f) for f in diff.get("added", [])],
        removed=[DiffFinding(**f) for f in diff.get("removed", [])],
        counts=diff.get("counts", {}),
    )


@router.post("/finding-changes/dismiss")
async def dismiss_finding_changes(
    sample_id: int = Query(..., description="Sample ID"),
) -> dict:
    """Dismiss the stored finding-level change diff for a sample."""
    from backend.analysis.finding_diff import dismiss_finding_diff
    from backend.api.routes.risk_common import resolve_sample_engine

    engine = resolve_sample_engine(sample_id)
    ok = dismiss_finding_diff(engine)
    if not ok:
        raise HTTPException(status_code=404, detail="No finding-change diff to dismiss.")
    return {"status": "dismissed", "sample_id": sample_id}


@router.post("/auto-update", response_model=AutoUpdateResponse)
async def toggle_auto_update(req: AutoUpdateRequest) -> AutoUpdateResponse:
    """Persist the per-database auto-update toggle.

    The set of valid ``db_name`` values is the union of the database
    registry and ``AUTO_UPDATE_DEFAULTS``. Unknown databases yield 404.
    """
    known = set(DATABASES) | set(AUTO_UPDATE_DEFAULTS)
    if req.db_name not in known:
        raise HTTPException(status_code=404, detail=f"Unknown database: {req.db_name}")

    registry = get_registry()
    set_auto_update(registry.reference_engine, req.db_name, req.enabled)
    return AutoUpdateResponse(db_name=req.db_name, enabled=req.enabled)

File: frontend/src/api/updates.ts
/** API hooks for database update status, history, and triggers (P4-17, P4-18). */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

// ── Types ────────────────────────────────────────────────────────────

export interface DatabaseStatus {
  db_name: string
  display_name: string
  current_version: string | null
  version_display: string | null
  downloaded_at: string | null
  file_size_bytes: number | null
  auto_update: boolean
  update_available: boolean
  /**
   * Bandwidth window for large updates (e.g. `"02:00-06:00"`) or `null`
   * when no window is configured. Mirrors `settings.update_download_window`
   * server-side; surfaced so the Update Manager can show an outside-window
   * tooltip on the "Update now" button. Optional — older backends omit it.
   */
  update_download_window?: string | null
}

export interface UpdateAvailable {
  db_name: string
  latest_version: string
  download_size_bytes: number
  release_date: string | null
}

export interface UpdateCheckResult {
  available: UpdateAvailable[]
  up_to_date: string[]
  errors: string[]
  checked_at: string
}

export interface UpdateHistoryEntry {
  id: number
  db_name: string
  previous_version: string | null
  new_version: string
  updated_at: string | null
  variants_added: number | null
  variants_reclassified: number | null
  download_size_bytes: number | null
  duration_seconds: number | null
}

export interface ReannotationPrompt {
  id: number
  sample_id: number
  db_name: string
  db_version: string
  candidate_count: number
  created_at: string | null
}

// ── Finding-level change diff (SW-A4b) ───────────────────────────────

export interface FindingFieldChange {
  field: string
  before: string | null
  after: string | null
}

export interface ReleaseDelta {
  db_name: string
  before: string | null
  after: string | null
}

export interface ChangedFinding {
  module: string
  category: string | null
  gene_symbol: string | null
  rsid: string | null
  drug: string | null
  diplotype: string | null
  finding_text: string
  changes: FindingFieldChange[]
}

export interface DiffFinding {
  module: string
  category: string | null
  gene_symbol: string | null
  rsid: string | null
  drug: string | null
  diplotype: string | null
  finding_text: string
  clinvar_significance: string | null
  evidence_level: number | null
  metabolizer_status: string | null
  pathway_level: string | null
}

export interface FindingChanges {
  available: boolean
  generated_at: string | null
  release_deltas: ReleaseDelta[]
  changed: ChangedFinding[]
  added: DiffFinding[]
  removed: DiffFinding[]
  counts: Record<string, number>
}

export interface TriggerUpdateResponse {
  job_id: string
  db_name: string
  message: string
}

export interface JobStatus {
  job_id: string
  status: string
  progress_pct: number
  message: string
  error: string | null
}

export interface AppUpdateInfo {
  update_available: boolean
  current_version: string
  latest_version: string | null
  release_url: string | null
  release_notes: string | null
  error: string | null
}

// ── Query keys ───────────────────────────────────────────────────────

export const DB_STATUS_KEY = ['updates', 'status'] as const
export const UPDATE_CHECK_KEY = ['updates', 'check'] as const
export const UPDATE_HISTORY_KEY = ['updates', 'history'] as const
export const REANNOTATION_PROMPTS_KEY = ['updates', 'prompts'] as const
export const APP_UPDATE_KEY = ['updates', 'app'] as const
export const FINDING_CHANGES_KEY = ['updates', 'finding-changes'] as const

// ── Fetchers ─────────────────────────────────────────────────────────

async function fetchDatabaseStatuses(): Promise<DatabaseStatus[]> {
  const res = await fetch('/api/updates/status')
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Database status fetch failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function fetchUpdateCheck(): Promise<UpdateCheckResult> {
  const res = await fetch('/api/updates/check')
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Update check failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function fetchUpdateHistory(dbName?: string, limit = 50): Promise<UpdateHistoryEntry[]> {
  const params = new URLSearchParams()
  if (dbName) params.set('db_name', dbName)
  params.set('limit', String(limit))
  const res = await fetch(`/api/updates/history?${params}`)
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Update history fetch failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function fetchReannotationPrompts(sampleId?: number): Promise<ReannotationPrompt[]> {
  const params = new URLSearchParams()
  if (sampleId != null) params.set('sample_id', String(sampleId))
  const res = await fetch(`/api/updates/prompts?${params}`)
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Prompts fetch failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function triggerUpdate(
  dbName: string,
  force = false,
): Promise<TriggerUpdateResponse> {
  // Only include `force` when truthy so legacy backends that don't know
  // the field still see the original payload shape.
  const body: Record<string, unknown> = { db_name: dbName }
  if (force) body.force = true
  const res = await fetch('/api/updates/trigger', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Trigger update failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function dismissPrompt(promptId: number): Promise<void> {
  const res = await fetch(`/api/updates/prompts/${promptId}/dismiss`, { method: 'POST' })
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Dismiss prompt failed: ${res.status} ${text}`.trim())
  }
}

async function fetchFindingChanges(sampleId: number): Promise<FindingChanges> {
  const res = await fetch(`/api/updates/finding-changes?sample_id=${sampleId}`)
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Finding-changes fetch failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function dismissFindingChanges(sampleId: number): Promise<void> {
  const res = await fetch(
    `/api/updates/finding-changes/dismiss?sample_id=${sampleId}`,
    { method: 'POST' },
  )
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Dismiss finding-changes failed: ${res.status} ${text}`.trim())
  }
}

async function toggleAutoUpdate(dbName: string, enabled: boolean): Promise<void> {
  const res = await fetch('/api/updates/auto-update', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ db_name: dbName, enabled }),
  })
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Toggle auto-update failed: ${res.status} ${text}`.trim())
  }
}

async function fetchJobStatus(jobId: string): Promise<JobStatus> {
  const res = await fetch(`/api/updates/job/${jobId}`)
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`Job status fetch failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

async function pollJobUntilDone(
  jobId: string,
  intervalMs = 2000,
  maxDurationMs = 10 * 60 * 1000, // 10 minutes
  signal?: AbortSignal,
): Promise<JobStatus> {
  const startTime = Date.now()
  while (true) {
    if (signal?.aborted) {
      throw new DOMException('Polling aborted', 'AbortError')
    }
    if (Date.now() - startTime > maxDurationMs) {
      throw new Error(`Job ${jobId} did not complete within ${maxDurationMs / 1000}s`)
    }
    const status = await fetchJobStatus(jobId)
    if (status.status === 'complete' || status.status === 'failed' || status.status === 'cancelled') {
      return status
    }
    await new Promise((r) => setTimeout(r, intervalMs))
  }
}

async function fetchAppUpdate(): Promise<AppUpdateInfo> {
  const res = await fetch('/api/updates/app-update')
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`App update check failed: ${res.status} ${text}`.trim())
  }
  return res.json()
}

// ── Hooks ────────────────────────────────────────────────────────────

/** Fetch per-DB version stamps and auto-update status. staleTime=1h. */
export function useDatabaseStatuses() {
  return useQuery({
    queryKey: DB_STATUS_KEY,
    queryFn: fetchDatabaseStatuses,
    staleTime: 60 * 60 * 1000, // 1 hour
  })
}

/** Check for app updates via GitHub Releases. staleTime=1h. */
export function useAppUpdate(enabled = true) {
  return useQuery({
    queryKey: APP_UPDATE_KEY,
    queryFn: fetchAppUpdate,
    staleTime: 60 * 60 * 1000, // 1 hour
    enabled,
    retry: false, // Don't retry on network failure — non-critical
  })
}

/** Check for available updates (hits remote). staleTime=1h. */
export function useUpdateCheck(enabled = false) {
  return useQuery({
    queryKey: UPDATE_CHECK_KEY,
    queryFn: fetchUpdateCheck,
    staleTime: 60 * 60 * 1000,
    enabled,
  })
}

/** Fetch update history, optionally filtered by db_name. */
export function useUpdateHistory(dbName?: string) {
  return useQuery({
    queryKey: [...UPDATE_HISTORY_KEY, dbName ?? 'all'],
    queryFn: () => fetchUpdateHistory(dbName),
    staleTime: 60 * 60 * 1000,
  })
}

/** Fetch active re-annotation prompts. */
export function useReannotationPrompts(sampleId?: number) {
  return useQuery({
    queryKey: [...REANNOTATION_PROMPTS_KEY, sampleId ?? 'all'],
    queryFn: () => fetchReannotationPrompts(sampleId),
    staleTime: 5 * 60 * 1000, // 5 min
  })
}

export interface TriggerUpdateVariables {
  dbName: string
  /** Bypass the bandwidth-window check (Force update). Off by default. */
  force?: boolean
}

/** Trigger a database update. Polls for job completion, then invalidates caches. */
export function useTriggerUpdate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ dbName, force = false }: TriggerUpdateVariables) => {
      const resp = await triggerUpdate(dbName, force)
      // Poll until the background Huey task finishes
      await pollJobUntilDone(resp.job_id)
      return resp
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: DB_STATUS_KEY })
      qc.invalidateQueries({ queryKey: UPDATE_CHECK_KEY })
      qc.invalidateQueries({ queryKey: UPDATE_HISTORY_KEY })
    },
  })
}

/** Dismiss a re-annotation prompt. Invalidates prompts cache on success. */
export function useDismissPrompt() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (promptId: number) => dismissPrompt(promptId),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: REANNOTATION_PROMPTS_KEY })
    },
  })
}

/** Fetch the finding-level change diff for a sample (SW-A4b). Disabled until a
 * sample id is known. */
export function useFindingChanges(sampleId: number | null | undefined) {
  return useQuery({
    queryKey: [...FINDING_CHANGES_KEY, sampleId ?? 'none'],
    queryFn: () => fetchFindingChanges(sampleId as number),
    enabled: sampleId != null,
    staleTime: 5 * 60 * 1000, // 5 min
  })
}

/** Dismiss a sample's stored finding-change diff. Invalidates the diff cache. */
export function useDismissFindingChanges() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (sampleId: number) => dismissFindingChanges(sampleId),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: FINDING_CHANGES_KEY })
    },
  })
}

/** Toggle auto-update for a database. Invalidates status cache on success. */
export function useToggleAutoUpdate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ dbName, enabled }: { dbName: string; enabled: boolean }) =>
      toggleAutoUpdate(dbName, enabled),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: DB_STATUS_KEY })
    },
  })
}
